spinalcordtoolbox/labels.py

```diff
@@ -238,6 +247,19 @@ def labelize_from_discs(img: Image, ref: Image) -> Image:
             if coordinates_ref[j + 1].z < z <= coordinates_ref[j].z:
                 out.data[int(x), int(y), int(z)] = coordinates_ref[j].value
 
+    # deal with the level above the top disc and the level below the bottom disc
+    for x, y, z, _ in coordinates_input:
+        for j in range(len(coordinates_ref) - 1):
```

Comment:
Ah wait, sorry. My logic might have been wrong. It seems like the `== 0` checks _are_ necessary to produce identical output. But I'm not quite sure why... :thinking:

Reply:
Ah, wait. I think I've made some sort of mistake in my suggestions. The new output doesn't match the old output. Let me try to figure out why, my apologies!

Reply:
EDIT: Ah! The reason my updated code doesn't work is actually because of a bug in the original code.

For the example data provided:

![image](https://user-images.githubusercontent.com/16181459/230156030-72ab7731-5bfc-4822-a5b9-aa95e419c978.png)

- `coordinates_ref[-1].z `is _not_ the top-most disc
- `coordinates_ref[0].z `is _not_ the bottom-most disc

So, for my suggestions to work correctly, we need to make sure to A) sort the coordinates or B) swap `0` and `-1` (or both?)

Reply:
Ah, but this actually reveals another subtle bug!! 

The original conditional says `< z <=` (i.e. it includes the above level, but doesn't include the below level).

So, this means that there's now a gap in between the bottom-most levels:

![image](https://user-images.githubusercontent.com/16181459/230156844-e66a444c-e55a-4497-acc7-8dcff1f91b35.png)

So we need to change the `z <` to `z <=`, which results in the correct image:

![image](https://user-images.githubusercontent.com/16181459/230157238-54faa6c5-ef09-4b66-bc54-8336a0248a5f.png)

With the final refactored loop being:

```
    # for all points in input, match the `z` coordinate to the appropriate vertebral level
    for x, y, z, _ in coordinates_input:
        # case 1: `z` is above the top-most disc label
        if z > coordinates_ref[0].z:
            out.data[int(x), int(y), int(z)] = coordinates_ref[0].value - 1
        # case 2: `z` is at or below the bottom-most disc label
        elif z <= coordinates_ref[-1].z:
            out.data[int(x), int(y), int(z)] = coordinates_ref[-1].value
        # case 3: `z` is between two disc labels, so find the correct vertebral level
        else:
            for j in range(len(coordinates_ref) - 1):
                if coordinates_ref[j + 1].z < z <= coordinates_ref[j].z:
                    out.data[int(x), int(y), int(z)] = coordinates_ref[j].value
```

Reply:
Thanks for this deep digging into the code! 💯 
I agree with your refactored loop. I updated it in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4072/commits/5e1fec5453d29387f7953838a6d0019da106656e